File: week-1/problem-set-recursion/prob3.py
"""Prob3 example.

This file belongs to the week 1 recursion practice solutions. It demonstrates
how recursion can reverse a string either by building a new string or by
swapping a mutable list in place.
"""


def reverse_string(s):
    """Return a reversed copy of the given string using recursion.

    Args:
        s (str): The s input used by this function.

    Returns:
        str: The string produced by the algorithm.
    """
    if len(s) <= 1:
        return s
    return reverse_string(s[1:]) + s[0]


def reverse_inplace(s, left=0, right=None):
    """Reverse a mutable sequence in place using recursion.

    Args:
        s (str): The s input used by this function.
        left (object): The left boundary or left-side data used by the algorithm.
        right (object): The right boundary or right-side data used by the algorithm.

    Returns:
        object: The value produced by this function.
    """
    if right is None:
        right = len(s) - 1

    if left >= right:
        return s

    s[left], s[right] = s[right], s[left]
    return reverse_inplace(s, left + 1, right - 1)


def main():
    """Run the small demonstration for this teaching example.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    text = "hello world"
    characters = list(text)

    print("Reversed copy:", reverse_string(text))
    reverse_inplace(characters)
    print("Reversed in place:", "".join(characters))


if __name__ == "__main__":
    main()

File: week-1/problem-set-recursion/prob3.cpp
/**
 * @file prob3.cpp
 * @brief C++ version of the recursive string-reversal example.
 */

#include <iostream>
#include <string>

using namespace std;

/**
 * @brief Return a reversed copy of the input string.
 * @param text const string& Original text to reverse.
 * @return string New string containing the characters of text in reverse order.
 */
string reverse_string(const string& text) {
    if (text.size() <= 1) {
        return text;
    }
    return reverse_string(text.substr(1)) + text[0];
}

/**
 * @brief Reverse a mutable string in place.
 * @param text string& String whose characters will be swapped in place.
 * @param left int Left index of the current recursive range.
 * @param right int Right index of the current recursive range.
 * @return string Updated string after the recursive swaps finish.
 */
string reverse_inplace(string& text, int left = 0, int right = -1) {
    if (right == -1) {
        right = static_cast<int>(text.size()) - 1;
    }
    if (left >= right) {
        return text;
    }

    swap(text[left], text[right]);
    return reverse_inplace(text, left + 1, right - 1);
}

/**
 * @brief Run the string-reversal demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    string text = "hello world";
    string characters = text;

    cout << "Reversed copy: " << reverse_string(text) << '\n';
    cout << "Reversed in place: " << reverse_inplace(characters) << '\n';
    return 0;
}
